Adding documetation for VSTest Integration

vagisha-nidhi opened the pull request:
Adding documetation for VSTest Integration. Laying out the details for the user experience and implementation.

contributes to https://github.com/tonerdo/coverlet/issues/395

MarcoRossignoli commented on documentation/VSTestIntegration.md:
```diff
@@ -0,0 +1,92 @@
+#### Default
+| Option | Summary |
+|-------------|------------------------------------|
+|Format              | Results format in which coverage output is generated. Default format is cobertura.| 
```
Should we provide default `CoverletOutput`?
With msbuild we are using `MSBuildProjectDirectory` https://github.com/tonerdo/coverlet/blob/0956d1e8da6b7265b0b5e4f93af3499c4be99cc8/src/coverlet.msbuild.tasks/coverlet.msbuild.props#L14 but there is some discussion in place https://github.com/tonerdo/coverlet/issues/152

vagisha-nidhi replied:
We will only be getting the coverage output from coverlet.core APIs and the responsibility of sending the file will be with the datacollector. 
By default, currently the files are put into the TestResults directory, Where this TestResults directory will be created is configurable via runsettings.

MarcoRossignoli commented on documentation/VSTestIntegration.md:
```diff
@@ -0,0 +1,92 @@
+
+| Entry point | How will code coverage be enabled? | Syntax                                                               |
+|-------------|------------------------------------|----------------------------------------------------------------------|
+|dotnet test CLI              | Through a switch to condition data collection | `dotnet test --collect:"XPlat Code Coverage"`   |
```
Is this real final name or only a placeholder?

vagisha-nidhi replied:
`XPlat Code Coverage` is the final name at present.

MarcoRossignoli commented on documentation/VSTestIntegration.md:
```diff
@@ -0,0 +1,92 @@
+<RunSettings>
+  <DataCollectionRunSettings>
+    <DataCollectors>
+      <DataCollector friendlyName="XPlat code coverage">
```
Tech question, test runner search for a class with `DataCollectorFriendlyName` attribute named `XPlat code coverage` from output build folder, right?

vagisha-nidhi replied:
It searches for the class in various folders including the output build folder and the nuget folders.
Basically we'll call a target before VSTest that would add the nuget path of the datacollectors to the search directories(it will be picked up directly from the nuget folder)

MarcoRossignoli commented on documentation/VSTestIntegration.md:
```diff
@@ -0,0 +1,92 @@
+
+
+#### Scope of Enhancement 
+Currently, advanced options are supported via runsettings. Providing support through additional command line arguments in vstest can be taken up separately.
```
> Providing support through additional command line arguments in vstest can be taken up separately.

Will be better user experience...but I think that it's require non trivial internal changes.

vagisha-nidhi replied:
Yes. This will include some internal vstest changes which we can pick up as an enhancement.

MarcoRossignoli commented on documentation/VSTestIntegration.md:
```diff
@@ -0,0 +1,92 @@
+## Implementation Details
+The proposed solution is implemented with the help of [datacollectors](https://github.com/Microsoft/vstest-docs/blob/master/docs/extensions/datacollector.md). 
+1. Outproc Datacollector : The outproc collector would always run in a separate process(datacollector.exe/datacollector.dll) than the process in which tests are being executed(testhost*.exe/testhost.dll). This datacollector would be responsible for calling into coverlet APIs for instrumenting dlls, collecting coverage results and sending the coverage output file back to test platform.
+2. Inproc Datacollector : The inproc collector in the testhost process executing the tests. This collector will be needed to remove the dependency on the exit handler to flush the hit files.
```
I don't see config to plug inproc datacollector...am I missing something?

vagisha-nidhi replied:
The inproc datacollector will be automatically plugged in with `--collect:"XPlat Code Coverage" `. We have test platform changes for this. The inproc datacollector will lie in the same nuget and get picked from there.

MarcoRossignoli replied:
ah ok thank's now it's more clear in/out behaviours of my past tests.

vagisha-nidhi commented:
> Workflow is enough clear to me at this point.
> Wait for @tonerdo thoughts/question.
> I could try to implement collectors with MS guys supervision if we agree on it.

@MarcoRossignoli Sure.
We are willing to contribute to your repo these datacollectors, We will be happy to do so!

MarcoRossignoli commented:
> We are willing to contribute to your repo these datacollectors, We will be happy to do so!

No problem to me!

tonerdo commented on documentation/VSTestIntegration.md:
```diff
@@ -0,0 +1,92 @@
+# Coverlet Integration with VSTest
+
+## Motivation
+The cross platform solution for Code Coverage generation for .NET Core projects(in a consistent manner).
```
```suggestion
The cross platform solution for Code Coverage generation for .NET Core projects (in a consistent manner).
```

MarcoRossignoli commented:
@tonerdo before go on with this and let @vagisha-nidhi start to work on collector we should try to merge https://github.com/tonerdo/coverlet/pull/393 so we'll use new contracts.

MarcoRossignoli commented:
@vagisha-nidhi can you capitalize folder name `Documentation` to better adhere to foundation repo guidelines https://github.com/dotnet/foundation/blob/master/guidance/repo-guide.md

vagisha-nidhi commented:
@MarcoRossignoli I have addressed the comments here.
Since #393 have contract changes in coverlet.core, it'll be good to prioritize it.
#392 is also related to these datacollector changes. Please look into this as well.

vagisha-nidhi commented:
@MarcoRossignoli @tonerdo It'll be really convenient for me if you could provide an ETA on [https://github.com/tonerdo/coverlet/pull/393](https://github.com/tonerdo/coverlet/pull/393)?

MarcoRossignoli commented:
Got some review from Toni https://github.com/tonerdo/coverlet/pull/393#issuecomment-491637617, I'll do my best to replace PR with simpler new one in a pair of days...I won't be available between Thursday-Sunday for a fast holiday.

MarcoRossignoli commented:
@tonerdo PTAL https://github.com/tonerdo/coverlet/pull/409

tonerdo commented:
@vagisha-nidhi #409 has been merged in

MarcoRossignoli commented:
@tonerdo we can merge this...if during developments something change we'll update

vagisha-nidhi commented:
@tonerdo This can be merged, right?
